### PR TITLE
Fixes the issue of not being able to stop scrolling by dragging the content again

### LINKED ISSUE #133529: [two_dimensional_scrollables] Scrolling doesn't stop on tap & Cannot accelerate scrolling on iOS
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

Scrolling behavior works as expected on web but not on IOS.

Run the [dartpad example ](https://dartpad.dev/?id=4424936c57ed13093eb389123383e894)on iOS

Scroll any direction, then try to stop the scrolling with a tap. It does not stop until it runs out of physics.

Scroll any direction, then try to scroll again to speed up or keep the scrolling going. It does not accelerate the scroll and you must wait for the physics to run out before you can scroll again.

### Expected results

Work just like any other scroll.

Where you can tap to halt/stop the scrolling. And be able to flick / scroll again to speed up the scrolling and keep the momentum.

### Actual results

Must wait for scrolling physics to stop until you can interact again.

### Code sample

<details><summary>Code sample</summary>

```dart
// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'dart:math' as math;

import 'package:flutter/foundation.dart';
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      scrollBehavior: const MaterialScrollBehavior().copyWith(
        // Mouse dragging enabled for this demo
        dragDevices: PointerDeviceKind.values.toSet(),
      ),
      debugShowCheckedModeBanner: false,
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatelessWidget {
  final String title;

  const MyHomePage({
    Key? key,
    required this.title,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(title)),
      body: TwoDimensionalGridView(
        diagonalDragBehavior: DiagonalDragBehavior.free,
        delegate: TwoDimensionalChildBuilderDelegate(
            maxXIndex: 9,
            maxYIndex: 9,
            builder: (BuildContext context, ChildVicinity vicinity) {
              return Container(
                color: vicinity.xIndex.isEven && vicinity.yIndex.isEven
                    ? Colors.amber[50]
                    : (vicinity.xIndex.isOdd && vicinity.yIndex.isOdd
                        ? Colors.purple[50]
                        : null),
                height: 200,
                width: 200,
                child: Center(
                    child: Text(
                        'Row ${vicinity.yIndex}: Column ${vicinity.xIndex}')),
              );
            }),
      ),
    );
  }
}

class TwoDimensionalGridView extends TwoDimensionalScrollView {
  const TwoDimensionalGridView({
    super.key,
    super.primary,
    super.mainAxis = Axis.vertical,
    super.verticalDetails = const ScrollableDetails.vertical(),
    super.horizontalDetails = const ScrollableDetails.horizontal(),
    required TwoDimensionalChildBuilderDelegate delegate,
    super.cacheExtent,
    super.diagonalDragBehavior = DiagonalDragBehavior.none,
    super.dragStartBehavior = DragStartBehavior.start,
    super.keyboardDismissBehavior = ScrollViewKeyboardDismissBehavior.manual,
    super.clipBehavior = Clip.hardEdge,
  }) : super(delegate: delegate);

  @override
  Widget buildViewport(
    BuildContext context,
    ViewportOffset verticalOffset,
    ViewportOffset horizontalOffset,
  ) {
    return TwoDimensionalGridViewport(
      horizontalOffset: horizontalOffset,
      horizontalAxisDirection: horizontalDetails.direction,
      verticalOffset: verticalOffset,
      verticalAxisDirection: verticalDetails.direction,
      mainAxis: mainAxis,
      delegate: delegate as TwoDimensionalChildBuilderDelegate,
      cacheExtent: cacheExtent,
      clipBehavior: clipBehavior,
    );
  }
}

class TwoDimensionalGridViewport extends TwoDimensionalViewport {
  const TwoDimensionalGridViewport({
    super.key,
    required super.verticalOffset,
    required super.verticalAxisDirection,
    required super.horizontalOffset,
    required super.horizontalAxisDirection,
    required TwoDimensionalChildBuilderDelegate super.delegate,
    required super.mainAxis,
    super.cacheExtent,
    super.clipBehavior = Clip.hardEdge,
  });

  @override
  RenderTwoDimensionalViewport createRenderObject(BuildContext context) {
    return RenderTwoDimensionalGridViewport(
      horizontalOffset: horizontalOffset,
      horizontalAxisDirection: horizontalAxisDirection,
      verticalOffset: verticalOffset,
      verticalAxisDirection: verticalAxisDirection,
      mainAxis: mainAxis,
      delegate: delegate as TwoDimensionalChildBuilderDelegate,
      childManager: context as TwoDimensionalChildManager,
      cacheExtent: cacheExtent,
      clipBehavior: clipBehavior,
    );
  }

  @override
  void updateRenderObject(
    BuildContext context,
    RenderTwoDimensionalGridViewport renderObject,
  ) {
    renderObject
      ..horizontalOffset = horizontalOffset
      ..horizontalAxisDirection = horizontalAxisDirection
      ..verticalOffset = verticalOffset
      ..verticalAxisDirection = verticalAxisDirection
      ..mainAxis = mainAxis
      ..delegate = delegate
      ..cacheExtent = cacheExtent
      ..clipBehavior = clipBehavior;
  }
}

class RenderTwoDimensionalGridViewport extends RenderTwoDimensionalViewport {
  RenderTwoDimensionalGridViewport({
    required super.horizontalOffset,
    required super.horizontalAxisDirection,
    required super.verticalOffset,
    required super.verticalAxisDirection,
    required TwoDimensionalChildBuilderDelegate delegate,
    required super.mainAxis,
    required super.childManager,
    super.cacheExtent,
    super.clipBehavior = Clip.hardEdge,
  }) : super(delegate: delegate);

  @override
  void layoutChildSequence() {
    final double horizontalPixels = horizontalOffset.pixels;
    final double verticalPixels = verticalOffset.pixels;
    final double viewportWidth = viewportDimension.width + cacheExtent;
    final double viewportHeight = viewportDimension.height + cacheExtent;
    final TwoDimensionalChildBuilderDelegate builderDelegate =
        delegate as TwoDimensionalChildBuilderDelegate;

    final int maxRowIndex = builderDelegate.maxYIndex!;
    final int maxColumnIndex = builderDelegate.maxXIndex!;

    final int leadingColumn = math.max((horizontalPixels / 200).floor(), 0);
    final int leadingRow = math.max((verticalPixels / 200).floor(), 0);
    final int trailingColumn = math.min(
      ((horizontalPixels + viewportWidth) / 200).ceil(),
      maxColumnIndex,
    );
    final int trailingRow = math.min(
      ((verticalPixels + viewportHeight) / 200).ceil(),
      maxRowIndex,
    );

    double xLayoutOffset = (leadingColumn * 200) - horizontalOffset.pixels;
    for (int column = leadingColumn; column <= trailingColumn; column++) {
      double yLayoutOffset = (leadingRow * 200) - verticalOffset.pixels;
      for (int row = leadingRow; row <= trailingRow; row++) {
        final ChildVicinity vicinity =
            ChildVicinity(xIndex: column, yIndex: row);
        final RenderBox child = buildOrObtainChildFor(vicinity)!;
        child.layout(constraints.tighten(width: 200.0, height: 200.0));

        // Subclasses only need to set the normalized layout offset. The super
        // class adjusts for reversed axes.
        parentDataOf(child).layoutOffset = Offset(xLayoutOffset, yLayoutOffset);
        yLayoutOffset += 200;
      }
      xLayoutOffset += 200;
    }

    // Set the min and max scroll extents for each axis.
    final double verticalExtent = 200 * (maxRowIndex + 1);
    verticalOffset.applyContentDimensions(
      0.0,
      clampDouble(
          verticalExtent - viewportDimension.height, 0.0, double.infinity),
    );
    final double horizontalExtent = 200 * (maxColumnIndex + 1);
    horizontalOffset.applyContentDimensions(
      0.0,
      clampDouble(
          horizontalExtent - viewportDimension.width, 0.0, double.infinity),
    );
    // Super class handles garbage collection too!
  }
}

```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

N/A

</details>

### Logs

<details><summary>Logs</summary>

```console
None
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
No errors
```

</details>

### PR DESCRIPTION
**Pull request by intonarumori**
Moves the `PanGestureRecognizer` used to drag the content along both axis to the outer `Scrollable` subclass.

- This solves the issue of the inner `Scrollable` gestures being disabled while the outer `Scrollable` is scrolling
- Enables the user to stop the scroll movement by dragging the content again

Fixes: https://github.com/flutter/flutter/issues/133529

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

### CODE REVIEW

**Review by Piinks — COMMENTED** [review]
Hi @intonarumori thanks for contributing! This change will need tests to verify that it works as expected. Thank you!

**Comment by intonarumori** [comment]
@Piinks Thanks for taking a look.

For tests my first idea was to look into the `CustomScrollView` tests and replicate them for the two dimensional version.
However, it looks like there is no concrete subclass of `TwoDimensionalScrollView` in the Flutter package, so I'm not that route would work.

https://github.com/flutter/flutter/blob/e7058f568faf49458f83d479577f66c923bf8445/packages/flutter/lib/src/widgets/scrollable.dart#L1693-L1695

If you can give me some hints of how to go about testing this I would be happy to look into it.

**Comment by Piinks** [comment]
For sure! There are some concrete classes just for testing, see the following for examples:

- https://github.com/flutter/flutter/blob/master/packages/flutter/test/widgets/two_dimensional_utils.dart
- https://github.com/flutter/flutter/blob/master/packages/flutter/test/widgets/two_dimensional_viewport_test.dart
- https://github.com/flutter/flutter/blob/master/packages/flutter/test/widgets/two_dimensional_scroll_view_test.dart

**Comment by jtkeyva** [comment]
@intonarumori @Piinks any movement on this? Such a critical feature/bug fix. Does this also allow to scroll again to keep momentum / accelerate like in a regular scrollview?

Thanks

File: packages/flutter/test/widgets/two_dimensional_scroll_view_test.dart [hunk 1]
```diff
@@ -323,5 +323,104 @@ void main() {
       expect(scrollable.widget.diagonalDragBehavior, DiagonalDragBehavior.weightedContinuous);
       expect(scrollable.widget.dragStartBehavior, DragStartBehavior.down);
     }, variant: TargetPlatformVariant.all());
+
+    testWidgets('Fling and tap to stop', (WidgetTester tester) async {
```
Comment (Piinks) [inline packages/flutter/test/widgets/two_dimensional_scroll_view_test.dart:327]: Can you update these tests to check two more things? These should verify they we are actually scrolling when these taps occur. Can you add the following:

- check the ScrollActivity.velocity in both axes when the test is about to tap and after.
  - the activity is available through the ScrollPosition
  - verifies we went from scrolling to stopping in response
- we should also verify the position is actually changing/stops changing as a result of flinging and then stopping it with a tap.
  - verify the ScrollPosition.pixels before flinging, before tapping, and after tapping

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 2]
```diff
@@ -2252,9 +2185,10 @@ class _HorizontalInnerDimensionState extends ScrollableState {
       case DiagonalDragBehavior.weightedEvent:
       case DiagonalDragBehavior.weightedContinuous:
       case DiagonalDragBehavior.free:
-        verticalScrollable._handleDragEnd(verticalDragDetails);
+        //verticalScrollable._handleDragEnd(verticalDragDetails);
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scrollable.dart:2188]: Should this be removed?

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 3]
```diff
@@ -2039,87 +2039,20 @@ class _VerticalOuterDimension extends Scrollable {
...
-    final ScrollableDetails details = ScrollableDetails(
-      direction: widget.axisDirection,
-      controller: _effectiveScrollController,
-      clipBehavior: widget.clipBehavior,
-    );
-    // Skip building a scrollbar here, the dual scrollbar is added in
-    // TwoDimensionalScrollableState.
-    return _configuration.buildOverscrollIndicator(context, child, details);
-  }
+class _TwoDimensionalDragForwarding {
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scrollable.dart:2042]: Public docs are not required for private classes (`///`), but can you add some docs to this for maintainers to better understand what it's for and where ti is used?

**Review by Piinks — COMMENTED** [review]
Thanks for your patience. I appreciate you taking on a complex fix!

**Review by Piinks — COMMENTED** [review]
Hey @intonarumori are you planning to return to this and address the feedback above?

**Comment by jtkeyva** [comment]
almost there!

**Comment by Piinks** [comment]
We have not heard back from @intonarumori on this, so I am going to close it and open a new PR so we can close this bug. Thanks for contributing!

**Comment by jtkeyva** [comment]
@Piinks sounds good thanks. oh hey, nice video btw on this very feature :)

**Comment by intonarumori** [comment]
> We have not heard back from @intonarumori on this, so I am going to close it and open a new PR so we can close this bug. Thanks for contributing!

@Piinks  Sorry for the late reply, I've been bogged down by work lately and didn't have the time to come back to this.
Many thanks for taking on this. Looking forward to learn from the new one and hopefully land something better formatted next time.

**Pull request closed without merging**